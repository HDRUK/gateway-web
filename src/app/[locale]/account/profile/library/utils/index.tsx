import { IconButton } from "@mui/material";
import { createColumnHelper } from "@tanstack/react-table";
import { LibraryListItem, SelectedLibrary } from "@/interfaces/Library";
import Link from "@/components/Link";
import StyledCheckbox from "@/components/StyledCheckbox";
import TooltipIcon from "@/components/TooltipIcon";
import { CheckIcon, DeleteForeverIcon } from "@/consts/icons";
import { RouteName } from "@/consts/routeName";

const columnHelper = createColumnHelper<LibraryListItem>();

const getColumns = ({
    handleSelect,
    handleRemove,
    selected,
    translations,
}: {
    handleSelect: (data: SelectedLibrary) => void;
    handleRemove: (id: string | number) => void;
    selected: SelectedLibrary;
    translations: { [id: string]: string };
}) => [
    columnHelper.display({
        id: "actions",
        meta: { isPinned: true },
        cell: ({ row }) => {
            const {
                id,
                datasetId,
                name,
                dataCustodianId,
                dataCustodian,
                darEnabled,
                darTemplatePublished,
                cohortEnabled,
            } = row.original;
            return (
                <div style={{ textAlign: "center" }}>
                    <StyledCheckbox
                        checked={selected[id]?.selected}
                        onChange={(_e, value) =>
                            handleSelect({
                                [id]: {
                                    selected: value,
                                    datasetId,
                                    name,
                                    teamId: dataCustodianId,
                                    teamName: dataCustodian,
                                    darEnabled,
                                    darTemplatePublished,
                                    cohortEnabled,
                                },
                            })
                        }
                        size="large"
                        sx={{ p: 0 }}
                        iconSx={{ mr: 0 }}
                    />
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),

    columnHelper.display({
        id: "name",
        cell: ({ row: { original } }) => {
            const { datasetId, name } = original;
            const linkHref = `/${RouteName.DATASET_ITEM}/${datasetId}`;

            return (
                <Link href={linkHref}>
                    <div style={{ textAlign: "center" }}>{name}</div>
                </Link>
            );
        },
        header: () => <span>{translations.name}</span>,
        size: 150,
    }),

    columnHelper.display({
        id: "darEnabled",
        cell: ({
            row: {
                original: { darEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {darEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => (
            <TooltipIcon
                buttonSx={{ p: 0 }}
                size="small"
                label={translations.darEnabled}
                content={translations.darEnabled}
            />
        ),
        size: 100,
    }),

    columnHelper.display({
        id: "cohortEnabled",
        cell: ({
            row: {
                original: { cohortEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {cohortEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => <span>{translations.cohortEnabled}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "dataCustodian",
        cell: ({
            row: {
                original: { dataCustodian },
            },
        }) => <div style={{ textAlign: "center" }}>{dataCustodian}</div>,
        header: () => <span>{translations.dataCustodian}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "entityType",
        cell: ({
            row: {
                original: { entityType },
            },
        }) => <div style={{ textAlign: "center" }}>{entityType}</div>,
        header: () => <span>{translations.entityType}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "delete",
        cell: ({ row: { original } }) => {
            return (
                <div style={{ textAlign: "center" }}>
                    <IconButton onClick={() => handleRemove(original.id)}>
                        <DeleteForeverIcon color="primary" />
                    </IconButton>
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),
];

export { getColumns };
