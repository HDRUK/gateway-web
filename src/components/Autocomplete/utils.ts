import { ValueType } from "./Autocomplete";

const getChipLabel = (
    options: { value: string | number; label: string }[],
    value: ValueType
) => {
    if (typeof value === "string") {
        return value;
    }

    return options.find(option => option.value === value)?.label;
};

export { getChipLabel };
