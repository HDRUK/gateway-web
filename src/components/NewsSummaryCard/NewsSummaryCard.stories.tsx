import type { Meta, StoryObj } from "@storybook/react";
import NewsSummaryCard from "@/components/NewsSummaryCard";

const meta: Meta<typeof NewsSummaryCard> = {
    component: NewsSummaryCard,
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof NewsSummaryCard>;

export const Default: Story = {
    args: {
        buttonText: "Read more",
        headline: "This is a headline",
        summary:
            "Cras ut sem eu ligula tincidunt aliquet. Lorem ipsum dolor sit amet. Nullam maximus risus et pharetra fringill. Suspendisse porttitor tortor et lectus pulvinar",
        date: "2023-01-01",
        url: "https://www.healthdatagateway.org/",
        imageLink: "/images/account/teams/integrations/create.jpg",
        imageAlt: "Image alt text",
    },
};
