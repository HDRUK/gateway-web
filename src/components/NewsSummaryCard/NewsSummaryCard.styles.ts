import { styled } from "@mui/material";

export const DateBox = styled("div")({
    textAlign: "center",
    color: "white",
    position: "absolute",
    top: 8,
    fontSize: 32,
    padding: 8,
    right: 14,
    background:
        "linear-gradient(141.65deg, #3DB28C 16.7%, #3DAE8D 29.38%, #3FA291 40.25%, #418F97 52.02%, #4474A0 62.89%, #475DA7 71.04%)",
});
