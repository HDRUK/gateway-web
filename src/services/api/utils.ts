import { NotificationOptions } from "@/interfaces/Api";
import { AxiosResponse } from "axios";
import { Error } from "@/interfaces/Error";
import notificationService from "../notification";

interface ErrorNotificationProps {
    props: NotificationOptions;
    errorResponse: AxiosResponse<Error>;
    method: "delete" | "post" | "put" | "get";
}

const errorNotification = ({
    errorResponse,
    method,
    props,
}: ErrorNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const { data, status } = errorResponse || {};

    const fallbackTitle = i18n.exists(`api:common.error.status.${status}`)
        ? t(`api:common.error.status.${status}`)
        : "There has been an error";

    const title = data?.title || fallbackTitle;
    const message =
        data?.message ||
        t(`api:common.error.${method}.message`, {
            item: props.itemName || "Item",
        });

    notificationService.apiError(message, {
        title,
        message,
        errors: data?.errors,
        ...notificationProps,
    });
};

interface SuccessNotificationProps {
    props: NotificationOptions;
    method: "delete" | "post" | "put";
}

const successNotification = ({ props, method }: SuccessNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const customMessage = `api:${props.localeKey}.success.${method}.message`;
    const shouldOverideMessage = i18n.exists(customMessage);

    const message = shouldOverideMessage
        ? t(customMessage)
        : t(`api:common.success.${method}.message`, {
              item: props.itemName || "Item",
          });

    notificationService.success(message, { ...notificationProps });
};

export { errorNotification, successNotification };
