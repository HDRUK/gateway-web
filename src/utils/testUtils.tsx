import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
