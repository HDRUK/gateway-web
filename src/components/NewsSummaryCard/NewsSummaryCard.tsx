import { Box, Button, CardActions } from "@mui/material";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import CardMedia from "@mui/material/CardMedia";
import Typography from "@mui/material/Typography";
import { ArrowForward } from "@/consts/icons";
import { formatDate } from "@/utils/date";
import EllipsisLineLimit from "../EllipsisLineLimit";
import Link from "../Link";
import { DateBox } from "./NewsSummaryCard.styles";

interface NewsSummaryCardProps {
    buttonText: string;
    headline: string;
    summary: string;
    date: string;
    url: string;
    imageLink?: string;
    imageAlt?: string;
    imageHeight?: string;
    variant?: "content" | "feature";
}

const NewsSummaryCard = ({
    variant = "content",
    buttonText,
    headline,
    summary,
    date,
    url,
    imageLink,
    imageAlt,
    imageHeight = "225px",
}: NewsSummaryCardProps) => {
    return (
        <Card
            sx={{
                width: "100%",
                maxWidth: { tablet: 345 },
                background: "transparent",
                position: "relative",
            }}>
            <CardMedia
                component="img"
                height={imageHeight}
                image={imageLink}
                alt={imageAlt}
            />
            <DateBox>
                <div>{formatDate(date, "MMM")}</div>
                <div>{formatDate(date, "DD")}</div>
            </DateBox>
            <CardContent>
                <Typography
                    gutterBottom
                    fontSize={20}
                    variant="h5"
                    component="div"
                    textAlign="left">
                    <EllipsisLineLimit maxLine={2} text={headline} />
                </Typography>
                <Typography variant="body2" textAlign="left" component="div">
                    <EllipsisLineLimit maxLine={4} text={summary} />
                </Typography>
                {variant === "content" && (
                    <Box sx={{ mt: 1 }}>
                        <Link href={url} underline="hover">
                            {buttonText}
                        </Link>
                    </Box>
                )}
            </CardContent>
            {variant === "feature" && (
                <CardActions>
                    <Link href={url} color="primary" passHref>
                        <Button
                            variant="text"
                            endIcon={<ArrowForward color="primary" />}>
                            {buttonText}
                        </Button>
                    </Link>
                </CardActions>
            )}
        </Card>
    );
};

export default NewsSummaryCard;
