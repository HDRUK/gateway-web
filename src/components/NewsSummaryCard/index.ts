import NewsSummaryCard from "./NewsSummaryCard";

export default NewsSummaryCard;
