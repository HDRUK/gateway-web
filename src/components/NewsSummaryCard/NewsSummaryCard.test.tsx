import React from "react";
import NewsSummaryCard from "@/components/NewsSummaryCard";
import { render } from "@/utils/testUtils";

describe("NewsSummaryCard", () => {
    it("should render component", async () => {
        const wrapper = render(
            <NewsSummaryCard
                buttonText="Read more"
                headline="This is a headline"
                summary="Cras ut sem eu ligula tincidunt aliquet. Lorem ipsum dolor sit amet. Nullam maximus risus et pharetra fringill. Suspendisse porttitor tortor et lectus pulvinar"
                date="2023-01-01"
                url="https://www.healthdatagateway.org/"
                imageLink="/images/account/teams/integrations/create.jpg"
                imageAlt="Image alt text"
            />
        );

        expect(wrapper.container).toMatchSnapshot();
    });
});
