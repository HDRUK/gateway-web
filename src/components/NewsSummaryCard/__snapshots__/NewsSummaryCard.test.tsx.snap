// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsSummaryCard should render component 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root mui-style-120eg3o-MuiPaper-root-MuiCard-root"
  >
    <img
      alt="Image alt text"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img mui-style-o69gx8-MuiCardMedia-root"
      height="225px"
      src="/images/account/teams/integrations/create.jpg"
    />
    <div
      class="mui-style-1sn3vv7"
    >
      <div>
        Jan
      </div>
      <div>
        01
      </div>
    </div>
    <div
      class="MuiCardContent-root mui-style-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom mui-style-96vc0a-MuiTypography-root"
      >
        <div
          class="LinesEllipsis  "
        >
          This is a headline
          <wbr />
        </div>
      </div>
      <div
        class="MuiTypography-root MuiTypography-body2 mui-style-qd21mq-MuiTypography-root"
      >
        <div
          class="LinesEllipsis  "
        >
          Cras ut sem eu ligula tincidunt aliquet. Lorem ipsum dolor sit amet. Nullam maximus risus et pharetra fringill. Suspendisse porttitor tortor et lectus pulvinar
          <wbr />
        </div>
      </div>
      <div
        class="MuiBox-root mui-style-164r41r"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover mui-style-zky3v8-MuiTypography-root-MuiLink-root"
          href="https://www.healthdatagateway.org/"
        >
          Read more
        </a>
      </div>
    </div>
  </div>
</div>
`;
